src/lib/ares_query.c

```diff
@@ -106,33 +106,13 @@ static void qcallback(void *arg, int status, int timeouts, unsigned char *abuf,
   if (status != ARES_SUCCESS) {
     qquery->callback(qquery->arg, status, timeouts, abuf, alen);
   } else {
-    /* Pull the response code and answer count from the packet. */
+    /* Pull the response code and answer count from the packet and convert any
+     * errors.
+     */
```

Comment:
Makes sense, I have switched to the new parsing functions.

On converting the status from the parent function:
- `ares_query` and `ares_search` both use `ares_send` for queries.
- I suppose the parent function calling the callback is `process_answer()`, which returns `ARES_SUCCESS` so long as we get any acceptable response from the server.
- The docs for `ares_send` mention `Unlike other query functions in the ares library, however, ares_send does not inspect the header of the reply packet to determine the error status, so a callback status of ARES_SUCCESS does not reflect as much about the response as for other query functions.`.

I think changing the parent function would be difficult without changing behaviour of `ares_send`. I'm not sure if that would be acceptable.

Reply:
you're right, that's a hard one.    My real goal is to prevent yet another parse of the DNS answer.

Maybe we should instead make a new `ares_send_dnsrec()` function that takes an `ares_dns_record_t` _and_ the callback instead of returning the response buffer returns the `ares_dns_record_t` since we already have this internally parsed in `ares_process.c`'s `process_answer()`.  This would prevent the need to double-parse, of course to hook into the 'legacy' functions it would need to do a `ares_dns_write()` to re-write the raw DNS packet.  This may be preferable anyhow since we can guarantee our rewritten packet is well-formed.

The new `ares_search_dnsrec()` you added should use this new same callback as well.  And probably all internal code should use it for efficiency to prevent the double-parse (however this can be delayed and doesn't need to be part of this PR).

That said, if you'd like to keep it how you have it and kick the can down the road, I'm ok with that, its something I might take on and tackle before the next release.

Reply:
Thanks, yep that makes sense. Especially in the case of an `ares_send_dnsrec()` function; if a user is already using the new DNS parser to create queries, they may want to use it in the callback too!

In terms of doing this here, I'm conscious that this PR is getting on the larger side. And I could imagine the further changes spreading across multiple places (need to design the API, rewrite `ares_send()`, add a new typedef for the callback, update the processing code). If possible I'd prefer to leave this PR limited to the `ares_search()` changes.

Is there anything I can do in this PR to help lay the foundation for the further changes? E.g. make sure we are not getting locked into the wrong API design before the next release?

Reply:
maybe just change the prototype of `ares_send_dnsrec()` to take a new callback type and pass an `ares_dns_record_t *` to that callback instead of the abuf/alen, even if that means you have to re-parse right now, then we can change the internals later.  I think this should be a const pointer, however `ares_dns_record_rr_get()` returns a writable pointer since you may be adding records, so we'd need a read-only variant of that (`ares_dns_record_rr_fetch()` ... `ares_dns_record_rr_get_const()` ... `ares_dns_record_rr_get_ro()` ??).  All the other fetchers take const pointers so its the standout.

Reply:
err, or maybe I mean just `ares_search_dnsrec()` for the prototype change since you haven't made `ares_send_dnsrec()` :)

Reply:
and for that new callback, should also make `int status` -> `ares_status_t status` and `int timeouts` -> `size_t timeouts`, to be more consistent with the latest coding standards for new APIs :)

Reply:
Thanks, I'm happy with that further change, it feels like it will future proof the API.

Here is a summary of the code changes I pushed:
- I added a new callback type `typedef void (*ares_callback_dnsrec)(void *arg, ares_status_t status, size_t timeouts,  const ares_dns_record_t *dnsrec);`. This is used by `ares_search_dnsrec()`.
- `ares_search_int()` continues to use the old `ares_callback` prototype (I thought it was better to optimise for the old `ares_search()`, at least until `ares_send_dnsrec()` is added).
  - To convert from `ares_callback->ares_callback_dnsrec`, there is a "conversion callback" named `ares__dnsrec_convert_cb()`. This parses the `abuf/alen` into an `ares_dns_record_t`.
  - When we eventually add `ares_send_dnsrec()`, we will probably want a "conversion callback" for the opposite direction.
- `ares_dns_record_rr_get_const()` actually already existed as a private function. I have exposed this externally.
- Whilst updating the regression tests, I noticed that `ares_dns_write()` doesn't take a `const` DNS record, but I think it should! I have fixed this (***note this changes the prototype in the API***).
- I added `SearchCallbackDnsRec()` to the regression tests to demo the new callback type.

Reply:
Do you think updating `ares_dns_write()` to take a `const ares_dns_record_t` is acceptable? Or would this need a new function `ares_dns_write_const()`?

If you are happy with the new callback, I will write up the man page docs for them.

Reply:
I have added the man page docs now. CI should be passing again too (I was blocked yesterday due to not having Cirrus CI credits, but that seems to be resolved).

Reply:
> I have added the man page docs now. CI should be passing again too (I was blocked yesterday due to not having Cirrus CI credits, but that seems to be resolved).

Yeah, I had to purchase some CI credits to reactivate it.  They only give a limited amount for open source projects like this and we burned through them pretty quick this month.  

I'll do a final pass review in a few days, I'm currently on a family vacation so my wife frowns on doing "work-like" things :)

Reply:
Ah OK, thanks for covering the CI; I will try to keep commits batched up, I know this PR has had a lot of iterations.

Thanks for taking the time to review the changes, enjoy your vacation!